Add avoid dependencies duplicates best practice (#32968)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/best-practices/bp_general.adoc

=== Tags
`#structuring-builds`

[[avoid_duplicate_dependencies]]

> == Avoid Declaring Dependencies Multiple Times

Reviewer: ```suggestion
== Avoid Redundant Dependency Declarations
```

> You should avoid duplicating dependencies in your build scripts, whether with different versions or in different dependency configurations.

Reviewer 2: I wonder if this is good _Gradle_ advice rather that good _general software development _ advice.

I think something like "don't duplicate already transitively present deps with direct declarations" would be more Gradle-specific, as such advice wouldn't necessarily apply to every build tool.

Reviewer 2: Also something like don't re-declare deps already present in extended configurations, like re-declaring something in `compileOnly` that's already in `implementation`.

I think you could probably cover all of these concerns in this single item.

> You should avoid duplicating dependencies in your build scripts, whether with different versions or in different dependency configurations.

Reviewer: Maybe something like this:

```suggestion
Avoid declaring the same dependency multiple times, especially when it is already available transitively or through another configuration.
```

Reviewer 3: I think this is specific to Gradle as Gradle will allow you to declare the same dependency on the same configuration or merged configurations without any warnings.

I think we should highlight that this is about redundant duplication *within* a single build.gradle.kts file, the way it is written it it makes it sound like this is about cross module, and I don't think we should tell people to remove explicit declarations between build.gradle.kts files.

Reviewer 2: I think this could also apply to declaring the same dep on confA and on confB which extendsFrom confA.

Author: > the way it is written it it makes it sound like this is about cross module, and I don't think we should tell people to remove explicit declarations between build.gradle.kts files.

Definitely, I'll update the part about transitivity. And I agree that this also applicable to configuration hierarchy.

=== Explanation

> Duplicating dependencies in build scripts can lead to several problems:

Reviewer: ```suggestion
Duplicating dependencies in Gradle build scripts can lead to:
```

> * Conflicting versions may cause unexpected behavior or runtime errors due to version mismatches

Reviewer 2: Like my comment below, conflicting versions are different dependencies.  Also, the example doesn't show a _conflict_, which is something that would fail a build, but merely a successful build that is kind of "irrational" in what it's requesting.

Author: >  Also, the example doesn't show a conflict, which is something that would fail a build, but merely a successful build that is kind of "irrational" in what it's requesting.

So you'd prefer to show the exact case when it would fail? e.g. display some code that works on version A and doesn't on version B?

Reviewer 2: I think the idea was to show multiple declarations for the same version, possibly one direct and one transitive - not 2 declarations for different versions that conflict.

So I was thinking of a case where the build _would not fail_, but the multiple declarations would still be unwanted.

Author: w.r.t transitive dependency and its direct duplicate. I think it is OK to have it.
Especially if user is explicit about it. 
e.g. I can have a ktor as a dependency. Which transitively brings some parts of kotlinx coroutines as API. And I'd like to use coroutines as well. So here I think it is logical to not rely on the fact that ktor exports it as well, but declare it explicitly.

And things got spicier when suddenly an upgrade of Ktor may upgrade my coroutines as well. But that is a topic of different BP.

So let me know if you're ok with the current example. feel free to resolve this thread if everything is ok.

Reviewer 2: I suppose it depends on the particular case.  I would say if your code needs a dep to compile, its okay to declare it directly, even if its already a transitive dep of another one of your deps.  But if you only need it because your direct deps require it, then you shouldn't declare it directly.

Reviewer 2: I think this example is okay now.

> * Conflicting versions may cause unexpected behavior or runtime errors due to version mismatches

Reviewer: A different version is not redundant, so we should remove this line.
```suggestion
```

> * Confusion during maintenance of build scripts and dependencies

Reviewer: ```suggestion
* Increased maintenance: Declaring a dependency in multiple places makes it harder to manage.
```

> 

Reviewer: ```suggestion
Unexpected behavior: Declaring the same dependency in multiple configurations (e.g., `compileOnly` and `implementation`) can result in hard-to-diagnose classpath issues.

```

=== Example

==== Don't Do This

====
include::sample[dir="snippets/bestPractices/avoidDuplicateDependencies/kotlin",files="build.gradle.kts[tags=avoid-this]"]
include::sample[dir="snippets/bestPractices/avoidDuplicateDependencies/groovy",files="build.gradle[tags=avoid-this]"]
====

<1> Dependency on library version `1.10.0`

> <2> Redundant declaration that can be removed

Reviewer 2: This isn't _redundant_, it's a request for a different version that gets replaced during resolution with the higher version.

This is actually an even _worse_ problem than _redundancy_.

Reviewer 2: Not declaring 2 different versions?  Yes, probably?

Author: Ok, I'll make it as separate one. and think about the good example.

=== Tags

`<<tags_reference.adoc#tag:upgrades,#upgrades>>`

> [[avoid_duplicate_dependencies]]

Reviewer 2: ❗ I don't think we want to change this file, the `bp_dependencies.adoc` change is all that is needed.

Author: yeah, this is mistake I overlooked. 
thank you

Author: fixed
